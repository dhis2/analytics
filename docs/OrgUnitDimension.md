# OrgUnitDimension Component

The OrgUnitDimension component is a part of the DHIS2 analytics package. It is used to select organisation units, levels, and groups.

## Props

- `roots`: An array of strings representing the roots of the organisation units.
- `selected`: An array of objects representing the selected organisation units. Each object should have an `id` and `name` property, and may optionally have a `path` property.
- `onSelect`: A function that is called when an organisation unit is selected.
- `hideGroupSelect`: A boolean that determines whether the group select should be hidden. Defaults to `false`.
- `hideLevelSelect`: A boolean that determines whether the level select should be hidden. Defaults to `false`.
- `hideUserOrgUnits`: A boolean that determines whether the user organisation units should be hidden. Defaults to `false`.
- `warning`: A string that represents a warning message.

## Usage
```javascript
import { OrgUnitDimension } from '@dhis2/analytics'

<OrgUnitDimension
    roots={['A0000001234']}
    selected={[{ id: 'B0000001234', name: 'Unit B' }]}
    onSelect={(selected) => console.log('Selected:', selected)}
    hideGroupSelect={false}
    hideLevelSelect={false}
    hideUserOrgUnits={false}
    warning="This is a warning message"
/>
```

