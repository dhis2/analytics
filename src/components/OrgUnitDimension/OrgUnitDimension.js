import React, { Component, Fragment } from 'react'
import PropTypes from 'prop-types'
import CircularProgress from '@material-ui/core/CircularProgress'
import DialogContent from '@material-ui/core/DialogContent'
import DialogTitle from '@material-ui/core/DialogTitle'
import sortBy from 'lodash/sortBy'

import i18n from '@dhis2/d2-i18n'
import { OrgUnitSelector, userOrgUnits } from '@dhis2/d2-ui-org-unit-dialog'

import {
    apiFetchOrganisationUnitGroups,
    apiFetchOrganisationUnitLevels,
    apiFetchOrganisationUnits,
} from '../../api/organisationUnits'

import { ouIdHelper } from '../../modules/ouIdHelper'

import { DIMENSION_ID_ORGUNIT } from '../../modules/fixedDimensions'
import styles from './styles/OrgUnitDimension.style'
import { colors } from '../../modules/colors'

export const defaultState = {
    root: undefined,
    roots: undefined,
    // use "selected" property for cloning org units while user org unit(s) is (are) selected
    selected: [],
    ouLevels: [],
    ouGroups: [],
    showOrgUnitsTree: true,
}

class OrgUnitDimension extends Component {
    constructor(props) {
        super(props)

        this.state = defaultState
        this.userOrgUnitIds = userOrgUnits.map(ou => ou.id)

        this.loadOrgUnitTree(props.d2, props.displayNameProperty)
        this.loadOrgUnitGroups(props.d2, props.displayNameProperty)
        this.loadOrgUnitLevels(props.d2)
    }

    componentDidUpdate(prevProps) {
        const previousId = prevProps.current ? prevProps.current.id : null
        const currentId = this.props.current ? this.props.current.id : null

        // remount org units selector component to ensure
        // only selected org units are expanded
        if (previousId !== currentId) {
            this.hideOrgUnitsTree()

            setTimeout(this.showOrgUnitsTree, 0)
        }
    }

    showOrgUnitsTree = () => {
        this.setState({ showOrgUnitsTree: true })
    }

    hideOrgUnitsTree = () => {
        this.setState({ showOrgUnitsTree: false })
    }

    getUserOrgUnitsFromIds = ids => {
        return userOrgUnits.filter(ou => ids.includes(ou.id))
    }

    onLevelChange = event => {
        const levelIds = event.target.value.filter(id => !!id)

        this.props.onSelect({
            dimensionId: DIMENSION_ID_ORGUNIT,
            items: [
                ...this.props.ouItems.filter(
                    ou => !ouIdHelper.hasLevelPrefix(ou.id)
                ),
                ...levelIds.map(id => {
                    const levelOu = this.state.ouLevels.find(ou => ou.id === id)

                    return {
                        ...levelOu,
                        id: ouIdHelper.addLevelPrefix(levelOu.id),
                    }
                }),
            ],
        })
    }

    onGroupChange = event => {
        const groupIds = event.target.value.filter(id => !!id)

        this.props.onSelect({
            dimensionId: DIMENSION_ID_ORGUNIT,
            items: [
                ...this.props.ouItems.filter(
                    ou => !ouIdHelper.hasGroupPrefix(ou.id)
                ),
                ...groupIds.map(id => {
                    const groupOu = this.state.ouGroups.find(ou => ou.id === id)

                    return {
                        ...groupOu,
                        id: ouIdHelper.addGroupPrefix(id),
                    }
                }),
            ],
        })
    }

    onDeselectAllClick = () =>
        this.props.onDeselect({
            dimensionId: DIMENSION_ID_ORGUNIT,
            itemIdsToRemove: this.props.ouItems.map(ou => ou.id),
        })

    loadOrgUnitTree = (d2, displayNameProperty) => {
        apiFetchOrganisationUnits(d2, displayNameProperty)
            .then(rootLevel => rootLevel.toArray())
            .then(roots => {
                this.setState({
                    roots,
                    root: roots[0],
                })
            })
    }

    loadOrgUnitGroups = (d2, displayNameProperty) => {
        apiFetchOrganisationUnitGroups(d2, displayNameProperty).then(
            organisationUnitGroups =>
                this.setState({ ouGroups: organisationUnitGroups })
        )
    }

    loadOrgUnitLevels = d2 => {
        apiFetchOrganisationUnitLevels(d2).then(organisationUnitLevels =>
            this.setState({
                ouLevels: sortBy(organisationUnitLevels, ['level']),
            })
        )
    }

    handleOrgUnitClick = (event, orgUnit) => {
        const selected = this.props.ouItems

        if (selected.some(ou => ou.path === orgUnit.path)) {
            this.props.onDeselect({
                dimensionId: DIMENSION_ID_ORGUNIT,
                itemIdsToRemove: [orgUnit.id],
            })
        } else {
            this.props.onSelect({
                dimensionId: DIMENSION_ID_ORGUNIT,
                items: [
                    ...selected,
                    {
                        ...orgUnit,
                        name: orgUnit.name || orgUnit.displayName,
                    },
                ],
            })
        }
    }

    handleUserOrgUnitClick = (event, checked) => {
        if (checked) {
            if (!this.state.selected.length) {
                this.setState({
                    selected: this.props.ouItems.slice(),
                })
            }

            this.props.onSelect({
                dimensionId: DIMENSION_ID_ORGUNIT,
                items: [
                    ...this.props.ouItems.filter(ou =>
                        this.userOrgUnitIds.includes(ou.id)
                    ),
                    userOrgUnits.find(ou => ou.id === event.target.name),
                ],
            })
        } else {
            if (
                this.props.ouItems.length === 1 &&
                this.state.selected.length > 0
            ) {
                this.props.onSelect({
                    dimensionId: DIMENSION_ID_ORGUNIT,
                    items: this.state.selected,
                })
            } else {
                this.props.onDeselect({
                    dimensionId: DIMENSION_ID_ORGUNIT,
                    itemIdsToRemove: [event.target.name],
                })
            }
        }
    }

    handleMultipleOrgUnitsSelect = orgUnits => {
        const selected = this.props.ouItems

        this.props.onSelect({
            dimensionId: DIMENSION_ID_ORGUNIT,
            items: [
                ...selected,
                ...orgUnits.reduce((obj, ou) => {
                    // avoid duplicates when clicking "Select children" multiple times
                    if (!selected.find(i => i.id === ou.id)) {
                        obj.push({ ...ou, name: ou.name || ou.displayName })
                    }

                    return obj
                }, []),
            ],
        })
    }

    render = () => {
        const ids = this.props.ouItems.map(ou => ou.id)
        const selected = this.props.ouItems.filter(
            ou =>
                !this.userOrgUnitIds.includes(ou.id) &&
                !ouIdHelper.hasLevelPrefix(ou.id) &&
                !ouIdHelper.hasGroupPrefix(ou.id)
        )

        const userOrgUnits = this.getUserOrgUnitsFromIds(ids)
        const level = ids
            .filter(ouIdHelper.hasLevelPrefix)
            .map(ouIdHelper.removePrefix)
        const group = ids
            .filter(ouIdHelper.hasGroupPrefix)
            .map(ouIdHelper.removePrefix)

        return (
            <Fragment>
                <DialogTitle>{i18n.t('Organisation units')}</DialogTitle>
                <DialogContent style={styles.dialogContent}>
                    {this.state.root && this.state.showOrgUnitsTree && (
                        <OrgUnitSelector
                            d2={this.props.d2}
                            root={this.state.root}
                            roots={this.state.roots}
                            selected={selected}
                            userOrgUnits={userOrgUnits}
                            level={level}
                            group={group}
                            levelOptions={this.state.ouLevels}
                            groupOptions={this.state.ouGroups}
                            onLevelChange={this.onLevelChange}
                            onGroupChange={this.onGroupChange}
                            onDeselectAllClick={this.onDeselectAllClick}
                            handleUserOrgUnitClick={this.handleUserOrgUnitClick}
                            handleOrgUnitClick={this.handleOrgUnitClick}
                            handleMultipleOrgUnitsSelect={
                                this.handleMultipleOrgUnitsSelect
                            }
                            checkboxColor="secondary"
                            deselectAllTooltipFontColor={colors.black}
                            deselectAllTooltipBackgroundColor={colors.greyLight}
                            displayNameProperty={this.props.displayNameProperty}
                            isUserDataViewFallback={true}
                        />
                    )}
                    {!this.state.root && (
                        <CircularProgress style={styles.loader} />
                    )}
                </DialogContent>
            </Fragment>
        )
    }
}

OrgUnitDimension.propTypes = {
    d2: PropTypes.object,
    displayNameProperty: PropTypes.string,
    onSelect: PropTypes.func,
    onDeselect: PropTypes.func,
    ouItems: PropTypes.array,
    current: PropTypes.object,
}

export default OrgUnitDimension
