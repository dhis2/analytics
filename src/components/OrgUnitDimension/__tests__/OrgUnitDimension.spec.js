import React from 'react'
import { shallow } from 'enzyme'

import OrgUnitDimension, { defaultState } from '../OrgUnitDimension'

/* eslint-disable react/display-name */
jest.mock('@dhis2/d2-ui-org-unit-dialog', () => {
    return {
        OrgUnitSelector: () => (
            <div id="mockOrgUnitSelector"> mockOrgUnitSelector</div>
        ),
        userOrgUnits: [
            {
                displayName: 'User organisation unit',
                id: 'USER_ORGUNIT',
            },
            {
                displayName: 'User sub-units',
                id: 'USER_ORGUNIT_CHILDREN',
            },
            {
                displayName: 'User sub-x2-units',
                id: 'USER_ORGUNIT_GRANDCHILDREN',
            },
        ],
        removeOrgUnitLastPathSegment: () => null,
    }
})
/* eslint-enable react/display-name */

jest.mock('../../../api/organisationUnits', () => {
    return {
        apiFetchOrganisationUnitGroups: () =>
            Promise.resolve([
                { displayName: 'CHC', id: 'CXw2yu5fodb', name: 'CHC' },
                {
                    displayName: 'Chiefdom',
                    id: 'gzcv65VyaGq',
                    name: 'Chiefdom',
                },
            ]),
        apiFetchOrganisationUnitLevels: () =>
            Promise.resolve([
                {
                    displayName: 'National',
                    id: 'H1KlN4QIauv',
                    level: 1,
                    name: 'National',
                },
                {
                    displayName: 'District',
                    id: 'wjP19dkFeIk',
                    level: 2,
                    name: 'District',
                },
            ]),
        apiFetchOrganisationUnits: () =>
            Promise.resolve({
                toArray: () => {
                    return [
                        {
                            id: 'jen',
                        },
                    ]
                },
            }),
    }
})

describe('The OrgUnitDimension component ', () => {
    let props
    let shallowDataDim

    const orgUnitDimension = () => {
        if (!shallowDataDim) {
            shallowDataDim = shallow(<OrgUnitDimension {...props} />)
        }
        return shallowDataDim
    }

    beforeEach(() => {
        props = {
            d2: {},
            ouItems: [],
            onSelect: jest.fn(),
            onDeselect: jest.fn(),
            current: { id: null },
            displayNameProperty: 'displayName',
        }
        shallowDataDim = undefined
    })

    it('has default state', () => {
        const actualState = orgUnitDimension().state()

        expect(actualState).toEqual(defaultState)
    })
})
