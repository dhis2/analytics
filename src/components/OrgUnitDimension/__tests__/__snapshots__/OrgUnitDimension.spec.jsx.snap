// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The OrgUnitDimension component matches the snapshot 1`] = `
<div
  className="container"
>
  <div
    className="userOrgUnitsWrapper"
  >
    <Checkbox
      checked={false}
      dataTest="dhis2-uicore-checkbox"
      dense={true}
      indeterminate={false}
      label="User organisation unit"
      onChange={[Function]}
    />
    <Checkbox
      checked={false}
      dataTest="dhis2-uicore-checkbox"
      dense={true}
      indeterminate={false}
      label="User sub-units"
      onChange={[Function]}
    />
    <Checkbox
      checked={false}
      dataTest="dhis2-uicore-checkbox"
      dense={true}
      indeterminate={false}
      label="User sub-x2-units"
      onChange={[Function]}
    />
  </div>
  <div
    className="orgUnitTreeWrapper"
  >
    <OrganisationUnitTree
      dataTest="org-unit-tree"
      initiallyExpanded={Array []}
      onChange={[Function]}
      roots={Array []}
      selected={Array []}
    />
  </div>
  <div
    className="selectsWrapper"
  >
    <MultiSelect
      dataTest="org-unit-level-select"
      dense={true}
      loading={true}
      onChange={[Function]}
      placeholder="Select a level"
      selected={Array []}
    />
    <MultiSelect
      dataTest="org-unit-group-select"
      dense={true}
      loading={true}
      onChange={[Function]}
      placeholder="Select a group"
      selected={Array []}
    />
  </div>
  <div
    className="summaryWrapper"
  >
    <span
      className="summaryText"
    >
      Nothing selected
    </span>
    <div
      className="deselectButton"
    >
      <Button
        disabled={true}
        onClick={[Function]}
        secondary={true}
        small={true}
      >
        Deselect all
      </Button>
    </div>
  </div>
  <style />
</div>
`;
